### PR TITLE
Heavy ion Vs algorithm same event subtraction

### PR DESCRIPTION
**Pull request by yslai**
A change to implement two features:

(a) A "fall-back" mode, when no calibration information is available, that estimates the UE from the same event using a close approximation of the ATLAS algorithm [Physics Letters B 719, 220–241 (2013)]. This mode is not turned on by default.

(b) Flags to expose the Fourier orders that are going to be subtracted. This was inadvertently hard-coded before.

The changes are merged due to otherwise conflicting diffs.

Issues raised previously from Carl Vuosalo's review are already addressed here.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoHI/HiJetAlgos/plugins/GenericVoronoiAlgorithm.cc [hunk 1]
```diff
@@ -0,0 +1,994 @@
...
+	// coefficient 0 + epsilon)
+	for (size_t i = _ncost; i <= index_column_max; i++) {
+		p->push_back_column(
+			epsilon_degeneracy * _recombine_tie[i - _ncost],
+			0, bpmpd_problem_t::infinity);
+	}
+}
+void GenericVoronoiAlgorithm::equalize(void)
+{
+	bpmpd_problem_t lp_problem = reinterpret_cast<bpmpd_environment_t *>(_lp_environment)->problem();
```
Comment (cvuosalo) [inline RecoHI/HiJetAlgos/plugins/GenericVoronoiAlgorithm.cc]: Where is _lp_environment set? I can't find any initialization for it.

File: RecoHI/HiJetAlgos/plugins/GenericVoronoiAlgorithm.h [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	double _positive_bound_scale;
+	std::vector<bool> _active;
+	std::vector<std::pair<size_t, size_t> > _recombine;
+	std::vector<std::vector<size_t> > _recombine_index;
+	std::vector<std::vector<size_t> > _recombine_unsigned;
+	std::vector<double> _recombine_tie;
+	size_t _ncost;
+	std::vector<size_t> _nblock_subtract;
+	void *_lp_environment;
+	void *_lp_problem;
```
Comment (cvuosalo) [inline RecoHI/HiJetAlgos/plugins/GenericVoronoiAlgorithm.h]: _lp_problem appears not to be used. What is it for?

File: RecoHI/HiJetAlgos/plugins/GenericVoronoiAlgorithm.cc (hunk 1 above)
Comment (yslai) [inline RecoHI/HiJetAlgos/plugins/GenericVoronoiAlgorithm.cc]: This is left over and removed together with the class (the previous class wrapper simply calls the default constructor). Fixed in the commit below.

File: RecoHI/HiJetAlgos/plugins/GenericVoronoiAlgorithm.h (hunk 2 above)
Comment (yslai) [inline RecoHI/HiJetAlgos/plugins/GenericVoronoiAlgorithm.h]: I missed this one when moving to a new PR, fixed also in the commit below.

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
@yslai:
There seem to be very large differences created by this PR. I tested workflow 140.53_RunHI2011 with 70 events against baseline CMSSW_7_6_X_2015-10-14-1600. Shown below are examples of large changes in jet energy and other values. What changes would you expect from this PR?
all_testpr11799vsorig_runhi2011wf140p53c_log10recopfjets_akvs4pfjets__rereco_obj_energy
all_testpr11799vsorig_runhi2011wf140p53c_log10recopfjets_akvs4pfjets__rereco_obj_et
all_testpr11799vsorig_runhi2011wf140p53c_log10recocalojets_akvs4calojets__rereco_obj_energy
all_testpr11799vsorig_runhi2011wf140p53c_recopfjets_akvs4pfjets__rereco_obj_chargedhadronenergyfraction
all_testpr11799vsorig_runhi2011wf140p53c_recopfjets_akvs4pfjets__rereco_obj_eta

**Comment by cvuosalo** [comment]
The static analyzer problems found in #11452 still have not been fixed. Please fix them now:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11799/8977/llvm-analysis/

**Comment by cvuosalo** [comment]
@yslai: When will fixes to this PR be ready? Time is running out for 76X.

**Comment by cvuosalo** [comment]
@yslai: When will the issues in this PR be addressed? If there is no progress in the near future, this PR may get rejected.

**Comment by yslai** [comment]
Due to a more urgent Vs bugfix, I am abandoning this PR, which will be resubmitted against the other PR and a later CMSSW version. My apologies.

**Pull request closed without merging**